collectors/ebpf.plugin/ebpf_hardirq.c

```diff
@@ -0,0 +1,503 @@
+        hardirq_read_latency_static_map(hardirq_maps[HARDIRQ_MAP_LATENCY_STATIC].map_fd);
+    }
+
+    read_thread_closed = 1;
+    return NULL;
+}
+
+static void hardirq_create_charts()
+{
+    ebpf_create_chart(
+        "system",
+        "hardirq_latency",
+        "Hardware IRQ latency",
+        "latency (milliseconds)",
+        "interrupts",
+        NULL,
+        NETDATA_EBPF_CHART_TYPE_STACKED,
+        NETDATA_CHART_PRIO_HARDIRQ_LATENCY,
+        NULL, NULL, 0,
+        NETDATA_EBPF_MODULE_NAME_HARDIRQ
+    );
+
+    fflush(stdout);
+}
+
+static void hardirq_create_static_dims()
+{
+    uint32_t i;
+    for (i = 0; i < HARDIRQ_EBPF_STATIC_END; i++) {
+        ebpf_write_global_dimension(
+            hardirq_static_vals[i].name, hardirq_static_vals[i].name,
+            ebpf_algorithms[NETDATA_EBPF_INCREMENTAL_IDX]
+        );
+    }
+}
+
+// callback for avl tree traversal on `hardirq_pub`.
+static int hardirq_write_dims(void *entry, void *data)
+{
+    UNUSED(data);
+
+    hardirq_val_t *v = entry;
+
+    // IRQs get dynamically added in, so add the dimension if we haven't yet.
+    if (!v->dim_exists) {
+        ebpf_write_global_dimension(
+            v->name, v->name,
+            ebpf_algorithms[NETDATA_EBPF_INCREMENTAL_IDX]
+        );
+        v->dim_exists = true;
+    }
+
+    write_chart_dimension(v->name, v->latency);
+
+    return 1;
+}
+
+static void hardirq_write_static_dims()
+{
+    uint32_t i;
+    for (i = 0; i < HARDIRQ_EBPF_STATIC_END; i++) {
+        write_chart_dimension(
+            hardirq_static_vals[i].name,
+            hardirq_static_vals[i].latency
+        );
+    }
+}
+
+/**
+* Main loop for this collector.
+*/
+static void hardirq_collector(ebpf_module_t *em)
+{
+    hardirq_ebpf_vals = callocz(
+        (running_on_kernel < NETDATA_KERNEL_V4_15) ? 1 : ebpf_nprocs,
+        sizeof(hardirq_ebpf_val_t)
+    );
+    hardirq_ebpf_static_vals = callocz(
+        (running_on_kernel < NETDATA_KERNEL_V4_15) ? 1 : ebpf_nprocs,
+        sizeof(hardirq_ebpf_static_val_t)
+    );
+
+    avl_init_lock(&hardirq_pub, hardirq_val_cmp);
+
+    // create reader thread.
+    hardirq_threads.thread = mallocz(sizeof(netdata_thread_t));
+    hardirq_threads.start_routine = hardirq_reader;
+    netdata_thread_create(
+        hardirq_threads.thread,
+        hardirq_threads.name,
+        NETDATA_THREAD_OPTION_JOINABLE,
+        hardirq_reader,
+        em
+    );
+
+    // create chart and static dims.
+    pthread_mutex_lock(&lock);
+    hardirq_create_charts();
+    hardirq_create_static_dims();
+    pthread_mutex_unlock(&lock);
+
+    // loop and read from published data until ebpf plugin is closed.
+    while (!close_ebpf_plugin) {
+        pthread_mutex_lock(&collect_data_mutex);
+        pthread_cond_wait(&collect_data_cond_var, &collect_data_mutex);
+        pthread_mutex_lock(&lock);
+
+        // write dims now for all hitherto discovered IRQs.
```

Comment:
A suggestion, the next four function calls are made to write charts for Netdata, I think they could go for a `static inline` function.
Like I said before, please, use the previous cited definitions instead to repeat the word "system" here.

Reply:
👍

Reply:
Just to be clear, the `write_begin_chart`, `write_end_chart` and `avl_traverse_lock` cannot be made `static inline` unless we moved them all to headers, since they come from other compilation units currently. So I assume you only meant to static inline `hardirq_write_static_dims`, which I'll do - let me know if I misunderstood.

Reply:
(Doing an inline on `hardirq_write_dims` is not going to do anything because it's being used as a callback at runtime).

Reply:
You are right, let us move it inside a specific function, I forgot where they were declared.